core/src/impl/Kokkos_ViewMapping.hpp

```diff
@@ -2982,22 +2982,24 @@ class ViewMapping<
   reference_type reference() const { return m_impl_handle[0]; }
 
   template <typename I0>
-  KOKKOS_FORCEINLINE_FUNCTION
-      typename std::enable_if<std::is_integral<I0>::value &&
-                                  !std::is_same<typename Traits::array_layout,
-                                                Kokkos::LayoutStride>::value,
-                              reference_type>::type
-      reference(const I0& i0) const {
+  KOKKOS_FORCEINLINE_FUNCTION typename std::enable_if<
+      std::is_integral<I0>::value &&
+          !std::is_same<typename Traits::array_layout,
+                        Kokkos::LayoutStride>::value &&
+          std::is_same<is_regular, std::true_type>::value,
+      reference_type>::type
+  reference(const I0& i0) const {
     return m_impl_handle[i0];
   }
 
   template <typename I0>
-  KOKKOS_FORCEINLINE_FUNCTION
-      typename std::enable_if<std::is_integral<I0>::value &&
-                                  std::is_same<typename Traits::array_layout,
-                                               Kokkos::LayoutStride>::value,
-                              reference_type>::type
-      reference(const I0& i0) const {
+  KOKKOS_FORCEINLINE_FUNCTION typename std::enable_if<
+      (std::is_integral<I0>::value &&
+       (std::is_same<typename Traits::array_layout,
+                     Kokkos::LayoutStride>::value ||
+        !std::is_same<is_regular, std::true_type>::value)),
```

Comment:
I prefer a direct negation of the same conditional above that it's providing an alternative to, so that it's clear to the reader that all of the cases are covered. e.g.,

```suggestion
       // comment highlighting that this is a negation of the above goes here
       !(
         // Helpful comment explaining this constraint goes here
         !std::is_same<typename Traits::array_layout,
                       Kokkos::LayoutStride>::value && is_regular::value),
```